### PR TITLE
core(offscreen-images): pass images with 'loading' attribute

### PR DESCRIPTION
**Pull request by jazyan**
**Summary**
Native lazy-loading was recently introduced in Chrome 76. This PR excludes images with a specified 'loading' attribute ('lazy' and 'eager', 'auto' TBD) from the 'defer offscreen images' check. 

See https://web.dev/native-lazy-loading/ for more details about the loading attribute. 

**Related Issues/PRs**
fixes: #10053

### CODE REVIEW

File: lighthouse-core/audits/byte-efficiency/offscreen-images.js [hunk 1]
```diff
@@ -76,7 +76,8 @@ class OffscreenImages extends ByteEfficiencyAudit {
    */
   static computeWaste(image, viewportDimensions, networkRecords) {
     const networkRecord = networkRecords.find(record => record.url === image.src);
-    if (!image.resourceSize || !networkRecord) {
+    if (!image.resourceSize || !networkRecord || image.loading === 'lazy' ||
+        image.loading === 'eager') {
```
Comment (jazyan) [inline lighthouse-core/audits/byte-efficiency/offscreen-images.js]: I tried adding `auto` in here as well, but since all images have `loading='auto'` by default, all the `FAIL(offscreen)` images in `smoke byte` were marked as passing. Open to suggestions!

File: lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js [hunk 2]
```diff
@@ -33,7 +33,7 @@ function generateSize(width, height, prefix = 'displayed') {
   return size;
 }
 
-function generateImage(size, coords, networkRecord, src = 'https://google.com/logo.png') {
+function generateImage(size, coords, networkRecord, src = 'https://google.com/logo.png', loading = null) {
```
Comment (jazyan) [inline lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js]: Not sure if `loading = null` is okay here, feels a bit weird.

Comment (connorjclark) [inline lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js]: maybe time to convert this to a function with a single object parameter, instead of 5 params. atm you must memorize the order of these params to understand each test case - would be much better as an object, like:

```js
generateImage({size: generateSize(x,y), x: coordx, y: coordy, ...})
```

File: lighthouse-core/audits/byte-efficiency/offscreen-images.js [hunk 3]
```diff
@@ -76,7 +76,8 @@ class OffscreenImages extends ByteEfficiencyAudit {
    */
   static computeWaste(image, viewportDimensions, networkRecords) {
     const networkRecord = networkRecords.find(record => record.url === image.src);
-    if (!image.resourceSize || !networkRecord) {
+    if (!image.resourceSize || !networkRecord || image.loading === 'lazy' ||
```
Comment (connorjclark) [inline lighthouse-core/audits/byte-efficiency/offscreen-images.js]: Are the first two conditions here error cases (why don't we have resourceSize or the network record), and the latter two legitimate cases? I find the mix to be odd, perhaps a comment would clear it up.

File: lighthouse-cli/test/fixtures/byte-efficiency/tester.html [hunk 4]
```diff
@@ -115,8 +115,8 @@ <h2>Byte efficiency tester page</h2>
   <!-- PASS(optimized): image is vector -->
   <!-- PASS(webp): image is vector -->
   <!-- PASS(responsive): image is vector -->
-  <!-- FAIL(offscreen): image is offscreen -->
```
Comment (connorjclark) [inline lighthouse-cli/test/fixtures/byte-efficiency/tester.html]: this seems to be the only failing case checking the `offscreen` audit in our smoke tests. Can we keep this?

File: lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js [hunk 5]
```diff
@@ -45,9 +45,14 @@ function generateImage(size, coords, networkRecord, src = 'https://google.com/lo
     left: x,
     right: x + size.displayedWidth,
   };
-  const image = {src, clientRect, ...networkRecord};
-  Object.assign(image, size);
-  return image;
+
+  return {
```
Comment (connorjclark) [inline lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js]: :100:

File: lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js [hunk 6]
```diff
@@ -128,6 +133,30 @@ describe('OffscreenImages audit', () => {
...
+    const url = s => `https://google.com/logo${s}.png`;
+    const topLevelTasks = [{ts: 1900, duration: 100}];
+    const networkRecords = [
+      generateRecord({url: url('A'), resourceSizeInKb: 100}),
+      generateRecord({url: url('B'), resourceSizeInKb: 100}),
+    ];
+    const artifacts = {
+      ViewportDimensions: DEFAULT_DIMENSIONS,
+      ImageElements: [
+        // offscreen to the right, lazy loaded
```
Comment (connorjclark) [inline lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js]: nit: we try to write our comments capitalized and ending with punctuation

File: lighthouse-core/audits/byte-efficiency/offscreen-images.js (hunk 3 above)
Comment (jazyan) [reply lighthouse-core/audits/byte-efficiency/offscreen-images.js]: Not totally sure if the first two conditions are error cases. Seems like we'll always have `resourceSize` but it [could be 0](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/gather/gatherers/image-elements.js#L208-L209), while the network record could be omitted for at least [this reason](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/gather/gatherers/image-elements.js#L171-L172), maybe others? LMK if that's accurate, I could write a comment explaining these cases.

File: lighthouse-cli/test/fixtures/byte-efficiency/tester.html [hunk 7]
```diff
@@ -38,7 +38,7 @@
 }
 
 // Add a long task to delay FI
-setTimeout(longTask, 2000);
```
Comment (patrickhulce) [inline lighthouse-cli/test/fixtures/byte-efficiency/tester.html]: why this change?

File: lighthouse-cli/test/fixtures/byte-efficiency/tester.html (hunk 4 above)
Comment (patrickhulce) [reply lighthouse-cli/test/fixtures/byte-efficiency/tester.html]: +1 to keeping this and adding a new, specific lazy loading images test image

File: lighthouse-core/audits/byte-efficiency/offscreen-images.js (hunk 3 above)
Comment (patrickhulce) [reply lighthouse-core/audits/byte-efficiency/offscreen-images.js]: First two cases are basically error cases, so I would agree with connor that we should separate and add a comment explaining. Something explaining why `loading="eager"` should pass would be great too.

```suggestion
    // If we don't know how big it was, we can't really report savings, treat it as passed.
    if (!image.resourceSize || !networkRecord) return null;
    // If the image had its loading behavior explicitly controlled already, treat it as passed.
    if (image.loading === 'lazy' || image.loading === 'eager') return null;
```

File: lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js [hunk 8]
```diff
@@ -109,16 +130,40 @@ describe('OffscreenImages audit', () => {
...
+        generateImage({
+          size: generateSize(100, 100),
+          coords: [0, 2000],
+          networkRecord: networkRecords[1],
+          src: url('B'),
+        }),
+        // Offscreen to the top-left.
+        generateImage({
+          size: generateSize(100, 100),
+          coords: [-2000, -1000],
```
Comment (connorjclark) [inline lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js]: wdyt of making `.coords` be `.x` and `.y`?

File: lighthouse-cli/test/fixtures/byte-efficiency/tester.html (hunk 7 above)
Comment (jazyan) [reply lighthouse-cli/test/fixtures/byte-efficiency/tester.html]: I noticed that when I originally added the `loading=lazy` test, it passed the `offscreen` audit even without my change. @paulirish helped me figure out why: it turned out that the image with native lazy loading was being filtered out [here](https://github.com/GoogleChrome/lighthouse/blob/master/lighthouse-core/audits/byte-efficiency/offscreen-images.js#L147). When we lengthened the below long task, the lazy loaded image request will load before TTI, so we were able to test the desired code path. Given that it depends on relative timing, it's not totally ideal, so would be open to learning if there's another way!

File: lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js (hunk 8 above)
Comment (jazyan) [reply lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js]: Sure! I don't have a particularly strong preference for `.coords`, so I can update it to `.x` and `.y`

File: lighthouse-cli/test/fixtures/byte-efficiency/tester.html (hunk 7 above)
Comment (patrickhulce) [reply lighthouse-cli/test/fixtures/byte-efficiency/tester.html]: Gotcha, that makes sense! There's not really a great way to isolate that in the smoketests without taking much, much longer than they do now unfortunately, so this will do! 👍

File: lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js [hunk 9]
```diff
@@ -128,13 +183,51 @@ describe('OffscreenImages audit', () => {
     assert.equal(auditResult.items.length, 4);
   });
 
+  it('passes images with a specified loading attribute', async () => {
```
Comment (exterkamp) [inline lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js]: nit: Check the opposite of this test to make sure that arbitrary loading attributes don't pass?

```javascript
it('fails images with a non-lazy loading attribute', async () => {
...
    const artifacts = {
      ViewportDimensions: DEFAULT_DIMENSIONS,
      ImageElements: [
          ...
          loading: 'not-lazy',
          ...
        }),
...
```

File: lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js [hunk 10]
```diff
@@ -221,6 +221,44 @@ describe('OffscreenImages audit', () => {
...
+          networkRecord: networkRecords[1],
+          loading: 'imagination',
+          src: url('B'),
+        }),
+      ],
+      traces: {defaultPass: createTestTrace({topLevelTasks})},
+      devtoolsLogs: {},
+    };
+
+    return UnusedImages.audit_(artifacts, networkRecords, context).then(auditResult => {
```
Comment (connorjclark) [inline lighthouse-core/test/audits/byte-efficiency/offscreen-images-test.js]: bit iffy on the `.then` in an `async` function, but all these other tests do the same so :man_shrugging:

**Comment by jazyan** [comment]
Thanks for the reviews! Merging is blocked for me, so if all looks good, would appreciate if an authorized user can merge this 😅

**Comment by connorjclark** [comment]
:tada:

**Comment by patrickhulce** [comment]
woohoo, great contribution @jazyan! 🎉
